Домашнее задание 22 спринта

DaniilKosmachev opened the pull request.

DaniilKosmachev committed:
ДЗ 22, этап 1
Сверстал экран добавления плейлиста
Реализована логика перехода на action

DaniilKosmachev committed:
ДЗ 22, этап 2
Реализована возможность выбора заставки
Реализована логика enabled для кнопки

DaniilKosmachev committed:
ДЗ 22, этап 3

DaniilKosmachev committed:
ДЗ 22, этап 4
Поправил сохранение в БД

DaniilKosmachev committed:
ДЗ 22, этап 5
Отображаются все плейлисты

DaniilKosmachev committed:
ДЗ 22, этап 6

DaniilKosmachev committed:
ДЗ 22, этап 7

DaniilKosmachev committed:
ДЗ 22, этап 8 (явно не последний)

Charmandik commented on app/src/main/java/com/example/playlistmaker/data/db/impl/PlaylistsRepositoryImpl.kt:
```diff
@@ -0,0 +1,47 @@
+
+    override suspend fun checkTrackInPlaylist(
+        trackId: Int
+    ): Flow<List<Int>> = flow {
```
В Kotlin лучше использовать "return" вместо "=", потому что это помогает делать код более читаемым и явным. Когда ты используешь "return", это явно указывает, что функция возвращает определенное значение. В то же время, использование "=" может вызывать путаницу, особенно если в функции есть несколько выражений.

Charmandik commented on app/src/main/java/com/example/playlistmaker/domain/player/model/TracksInPlaylistsState.kt:
```diff
@@ -0,0 +1,10 @@
+
+sealed interface TracksInPlaylistsState {
+    data class inPlaylist(
+        var data: List<TracksInPlaylists>
```
Использование var внутри data класса противоречит его смыслу, так как в идеале он должен быть полностью неизменяемым. Для ситуаций, когда нужно поменять значение внутри такого класса лучше создать новый экземпляр этого же класса, используя функцию .copy(), которая есть в каждом data классе

Charmandik commented on app/src/main/java/com/example/playlistmaker/ui/audioplayer/fragment/AudioPlayerFragment.kt:
```diff
@@ -0,0 +1,264 @@
+    private var _binding: FragmentAudioPlayerBinding? = null
+    private val binding get() = _binding!!
+
+    private lateinit var selectableTrack: Track
```
Вместо использования lateinit, я рекомендую тебе использовать подход с добавлением nullability к типу объекта с оператором ?. и сразу проинициализировать его значением null. Это решение позволяет избежать ошибок, связанных с попытками доступа к неинициализированным объектам, которые могут привести к непредсказуемому поведению или крашам приложения. Хотя это может добавить некоторую дополнительную работу при обращении к объекту через оператор ?. , но это значительно повысит безопасность приложения, так как ты будешь уверен, что объект всегда либо имеет значение, либо равен null, и не будет вызывать исключительных ситуаций.Можешь еще прочитать статью на эту тему https://habr.com/ru/companies/maxilect/articles/590061

Charmandik commented on app/src/main/java/com/example/playlistmaker/ui/audioplayer/fragment/AudioPlayerFragment.kt:
```diff
@@ -0,0 +1,264 @@
+            override fun onStateChanged(bottomSheet: View, newState: Int) {
+                when (newState) {
+                    BottomSheetBehavior.STATE_HIDDEN -> {
+                        binding.overlayBottomSheet.visibility = View.GONE
```
Для управления видимостью можно использовать свойство isVisible, устанавливая значение false для View.GONE и true для View.VISIBLE
Например, если нам нужно установить View.VISIBLE:
myViewName.isVisible = true
Или если нужно установить View.GONE, то : 
myViewName.isVisible = false

Возможно, тебе больше понравится такой подход

Charmandik commented on app/src/main/java/com/example/playlistmaker/ui/audioplayer/view_model/PlayerFragmentViewModel.kt:
```diff
@@ -31,6 +33,14 @@ class PlayerActivityViewModel(
 
+    private var mutablePlaylistStatus = MutableLiveData<PlaylistsState>()
+
+    private var mutableStatusCheckTrackInPlaylist = MutableLiveData<List<Int>>()
```
Для ситуаций, когда нам нужно иметь два состояния( мутабельное и иммутабельное), рекомендую использовать подход с добавлением префикса в виде нижнего подчеркивания для мутабельного состояния(которое должно быть приватным):
private val _state = MutableLiveData<MyState>()
val state : LiveData<MyState> = _state

Так ты сможешь не придумывать разные имена, чтобы не повторяться, но и при этом нормально передавать смысл и сможешь точно понимать, что эта вот переменная это просто копия той, но иммутабельна

DaniilKosmachev committed:
ДЗ 22, этап 9
Исправлены некритические замечания
Реализованы plurals для строк со словом трек
